The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15686860339/job/44192227912) [(plain)](https://github.com/rust-lang/rust/commit/55b745edaa7fbac3cbad51a24f60c5979a07c7b9/checks/44192227912/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
1 error[E0597]: `l` does not live long enough
-   --> $DIR/span-semicolon-issue-139049.rs:11:41
+   --> $DIR/span-semicolon-issue-139049.rs:13:11
3    |
- LL | macro_rules! perform { ($e:expr) => { D(&$e).end() } }
-    |                                       --^^^-
-    |                                       | |
-    |                                       | borrowed value does not live long enough
-    |                                       a temporary with access to the borrow is created here ...
+ LL |         D(&$e).end()
+    |         --^^^-
+    |         | |
+    |         | borrowed value does not live long enough
+    |         a temporary with access to the borrow is created here ...
9 ...
- LL |     { let l = (); perform!(l) }; 
-    |           -       ----------- -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
-    |           |       |           |
-    |           |       |           `l` dropped here while still borrowed
-    |           |       in this macro invocation
-    |           binding `l` declared here
+ LL |         let l = ();
+    |             - binding `l` declared here
+ LL |         perform!(l)
+    |         ----------- in this macro invocation
+ LL |     };
+    |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
+    |     |
+    |     `l` dropped here while still borrowed
16    |
17    = note: this error originates in the macro `perform` (in Nightly builds, run with -Z macro-backtrace for more info)
18 help: consider adding semicolon after the expression so its temporaries are dropped sooner, before the local variables declared by the block are dropped

19    |
- LL |     { let l = (); perform!(l); }; 
-    |                              +
+ LL |         perform!(l);
+    |                    +
22 
23 error[E0597]: `l` does not live long enough
-   --> $DIR/span-semicolon-issue-139049.rs:11:41
+   --> $DIR/span-semicolon-issue-139049.rs:13:11
25    |
- LL | macro_rules! perform { ($e:expr) => { D(&$e).end() } }
-    |                                       --^^^-
-    |                                       | |
-    |                                       | borrowed value does not live long enough
-    |                                       a temporary with access to the borrow is created here ...
+ LL |         D(&$e).end()
+    |         --^^^-
+    |         | |
+    |         | borrowed value does not live long enough
+    |         a temporary with access to the borrow is created here ...
31 ...
- LL |     let _x = { let l = (); perform!(l) };
-    |                    -       ----------- -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
-    |                    |       |           |
-    |                    |       |           `l` dropped here while still borrowed
-    |                    |       in this macro invocation
-    |                    binding `l` declared here
+ LL |         let l = ();
+    |             - binding `l` declared here
+ LL |         perform!(l)
+    |         ----------- in this macro invocation
+ LL |     };
+    |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
+    |     |
+    |     `l` dropped here while still borrowed
38    |
39    = note: the temporary is part of an expression at the end of a block;
40            consider forcing this temporary to be dropped sooner, before the block's local variables are dropped

41    = note: this error originates in the macro `perform` (in Nightly builds, run with -Z macro-backtrace for more info)
42 help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
43    |
- LL |     let _x = { let l = (); let x = perform!(l); x };
-    |                            +++++++            +++
+ LL |         let x = perform!(l); x
+    |         +++++++            +++
46 
47 error: aborting due to 2 previous errors
48 

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/borrowck/span-semicolon-issue-139049.rs:13:11
-   --> /checkout/tests/ui/borrowck/span-semicolon-issue-139049.rs:13:11
+   --> $DIR/span-semicolon-issue-139049.rs:13:11
+ LL |         D(&$e).end()
+    |         --^^^-
+    |         | |
+    |         | borrowed value does not live long enough
+    |         a temporary with access to the borrow is created here ...
+ LL |         let l = ();
+    |             - binding `l` declared here
+ LL |         perform!(l)
+    |         ----------- in this macro invocation
+ LL |     };
+    |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
+    |     |
+    |     `l` dropped here while still borrowed
+ LL |         perform!(l);
+    |                    +
+   --> $DIR/span-semicolon-issue-139049.rs:13:11
+ LL |         D(&$e).end()
+    |         --^^^-
+    |         | |
+    |         | borrowed value does not live long enough
+    |         a temporary with access to the borrow is created here ...
+ LL |         let l = ();
+    |             - binding `l` declared here
+ LL |         perform!(l)
+    |         ----------- in this macro invocation
+ LL |     };
+    |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
+    |     |
+    |     `l` dropped here while still borrowed
+ LL |         let x = perform!(l); x
+    |         +++++++            +++


The actual stderr differed from the expected stderr
Saved the actual fixed to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/borrowck/span-semicolon-issue-139049/span-semicolon-issue-139049.fixed`
diff of fixed:

8 
9 // You could assume that this comes from an extern crate (it doesn't
10 // because an aux crate would be overkill for this test).
- macro_rules! perform { ($e:expr) => { D(&$e).end() } }
+ macro_rules! perform {
+     ($e:expr) => {
+         D(&$e).end()
+     };
+ }
12 //~^ ERROR does not live long enough
13 //~| ERROR does not live long enough
14 

15 fn main() {
-     { let l = (); perform!(l); }; 
+     {
+         let l = ();
+         perform!(l);
+     };
17     //~^ SUGGESTION ;
-     let _x = { let l = (); let x = perform!(l); x };
+     let _x = {
+         let l = ();
+         let x = perform!(l); x
+     };
19     //~^ SUGGESTION let x
20 }
21 

22 struct D<T>(T);
- impl<T> Drop for D<T> { fn drop(&mut self) {} }
- impl<T> D<T> { fn end(&self) -> String { String::new()} }
+ impl<T> Drop for D<T> {
+     fn drop(&mut self) {}
+ }
+ impl<T> D<T> {
+     fn end(&self) -> String {
+         String::new()
+     }
---
To only update this specific test, also pass `--test-args borrowck/span-semicolon-issue-139049.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/borrowck/span-semicolon-issue-139049.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/borrowck/span-semicolon-issue-139049" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0597]: `l` does not live long enough
##[error]  --> /checkout/tests/ui/borrowck/span-semicolon-issue-139049.rs:13:11
   |
LL |         D(&$e).end()
   |         --^^^-
   |         | |
   |         | borrowed value does not live long enough
   |         a temporary with access to the borrow is created here ...
...
LL |         let l = ();
   |             - binding `l` declared here
LL |         perform!(l)
   |         ----------- in this macro invocation
LL |     };
   |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
   |     |
   |     `l` dropped here while still borrowed
   |
   = note: this error originates in the macro `perform` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider adding semicolon after the expression so its temporaries are dropped sooner, before the local variables declared by the block are dropped
   |
LL |         perform!(l);
   |                    +

error[E0597]: `l` does not live long enough
##[error]  --> /checkout/tests/ui/borrowck/span-semicolon-issue-139049.rs:13:11
   |
LL |         D(&$e).end()
   |         --^^^-
   |         | |
   |         | borrowed value does not live long enough
   |         a temporary with access to the borrow is created here ...
...
LL |         let l = ();
   |             - binding `l` declared here
LL |         perform!(l)
   |         ----------- in this macro invocation
LL |     };
   |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
   |     |
   |     `l` dropped here while still borrowed
   |
   = note: the temporary is part of an expression at the end of a block;
           consider forcing this temporary to be dropped sooner, before the block's local variables are dropped
   = note: this error originates in the macro `perform` (in Nightly builds, run with -Z macro-backtrace for more info)
help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
   |
LL |         let x = perform!(l); x
   |         +++++++            +++

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
```

</details>

--- a/compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs
@@ -342,10 +342,7 @@ impl<'tcx> BorrowExplanation<'tcx> {
                                 }
                             }
                         } else if let LocalInfo::BlockTailTemp(info) = local_decl.local_info() {
-                            let sp = info
-                                .span
-                                .find_ancestor_in_same_ctxt(local_decl.source_info.span)
-                                .unwrap_or(info.span);
+                            let sp = info.span.find_oldest_ancestor_in_same_ctxt();
                             if info.tail_result_is_ignored {
                                 // #85581: If the first mutable borrow's scope contains
                                 // the second borrow, this suggestion isn't helpful.

--- a/tests/ui/borrowck/span-semicolon-issue-139049.fixed
+++ b/tests/ui/borrowck/span-semicolon-issue-139049.fixed
@@ -1,52 +1,25 @@
-// Make sure the generated suggestion suggest editing the user
-// code instead of the std macro implementation
+// Make sure the generated suggestion suggest editing the user code instead of
+// the macro implementation (which might come from an external crate).
+// issue: <https://github.com/rust-lang/rust/issues/139049>
 
 //@ run-rustfix
 
 #![allow(dead_code)]
 
-use std::fmt::{self, Display};
-
-struct Mutex;
-
-impl Mutex {
-    fn lock(&self) -> MutexGuard<'_> {
-        MutexGuard(self)
-    }
-}
-
-struct MutexGuard<'a>(&'a Mutex);
-
-impl<'a> Drop for MutexGuard<'a> {
-    fn drop(&mut self) {}
-}
-
-struct Out;
-
-impl Out {
-    fn write_fmt(&self, _args: fmt::Arguments) {}
-}
-
-impl<'a> Display for MutexGuard<'a> {
-    fn fmt(&self, _formatter: &mut fmt::Formatter) -> fmt::Result {
-        Ok(())
-    }
-}
+// You could assume that this comes from an extern crate (it doesn't
+// because an aux crate would be overkill for this test).
+macro_rules! perform { ($e:expr) => { D(&$e).end() } }
+//~^ ERROR does not live long enough
+//~| ERROR does not live long enough
 
 fn main() {
-    let _write = {
-        let mutex = Mutex;
-        write!(Out, "{}", mutex.lock());
-        //~^ ERROR `mutex` does not live long enough
-        //~| SUGGESTION ;
-    };
+    { let l = (); perform!(l); };
+    //~^ SUGGESTION ;
 
-    let _write = {
-        use std::io::Write as _;
-
-        let mutex = Mutex;
-        let x = write!(std::io::stdout(), "{}", mutex.lock()); x
-        //~^ ERROR `mutex` does not live long enough
-        //~| SUGGESTION let x
-    };
+    let _x = { let l = (); let x = perform!(l); x };
+    //~^ SUGGESTION let x
 }
+
+struct D<T>(T);
+impl<T> Drop for D<T> { fn drop(&mut self) {} }
+impl<T> D<T> { fn end(&self) -> String { String::new() } }

--- a/tests/ui/borrowck/span-semicolon-issue-139049.rs
+++ b/tests/ui/borrowck/span-semicolon-issue-139049.rs
@@ -1,52 +1,25 @@
-// Make sure the generated suggestion suggest editing the user
-// code instead of the std macro implementation
+// Make sure the generated suggestion suggest editing the user code instead of
+// the macro implementation (which might come from an external crate).
+// issue: <https://github.com/rust-lang/rust/issues/139049>
 
 //@ run-rustfix
 
 #![allow(dead_code)]
 
-use std::fmt::{self, Display};
-
-struct Mutex;
-
-impl Mutex {
-    fn lock(&self) -> MutexGuard<'_> {
-        MutexGuard(self)
-    }
-}
-
-struct MutexGuard<'a>(&'a Mutex);
-
-impl<'a> Drop for MutexGuard<'a> {
-    fn drop(&mut self) {}
-}
-
-struct Out;
-
-impl Out {
-    fn write_fmt(&self, _args: fmt::Arguments) {}
-}
-
-impl<'a> Display for MutexGuard<'a> {
-    fn fmt(&self, _formatter: &mut fmt::Formatter) -> fmt::Result {
-        Ok(())
-    }
-}
+// You could assume that this comes from an extern crate (it doesn't
+// because an aux crate would be overkill for this test).
+macro_rules! perform { ($e:expr) => { D(&$e).end() } }
+//~^ ERROR does not live long enough
+//~| ERROR does not live long enough
 
 fn main() {
-    let _write = {
-        let mutex = Mutex;
-        write!(Out, "{}", mutex.lock())
-        //~^ ERROR `mutex` does not live long enough
-        //~| SUGGESTION ;
-    };
+    { let l = (); perform!(l) };
+    //~^ SUGGESTION ;
 
-    let _write = {
-        use std::io::Write as _;
-
-        let mutex = Mutex;
-        write!(std::io::stdout(), "{}", mutex.lock())
-        //~^ ERROR `mutex` does not live long enough
-        //~| SUGGESTION let x
-    };
+    let _x = { let l = (); perform!(l) };
+    //~^ SUGGESTION let x
 }
+
+struct D<T>(T);
+impl<T> Drop for D<T> { fn drop(&mut self) {} }
+impl<T> D<T> { fn end(&self) -> String { String::new() } }

--- a/tests/ui/borrowck/span-semicolon-issue-139049.stderr
+++ b/tests/ui/borrowck/span-semicolon-issue-139049.stderr
@@ -1,46 +1,48 @@
-error[E0597]: `mutex` does not live long enough
-  --> $DIR/span-semicolon-issue-139049.rs:39:27
+error[E0597]: `l` does not live long enough
+  --> $DIR/span-semicolon-issue-139049.rs:11:41
    |
-LL |         let mutex = Mutex;
-   |             ----- binding `mutex` declared here
-LL |         write!(Out, "{}", mutex.lock())
-   |                           ^^^^^-------
-   |                           |
-   |                           borrowed value does not live long enough
-   |                           a temporary with access to the borrow is created here ...
+LL | macro_rules! perform { ($e:expr) => { D(&$e).end() } }
+   |                                       --^^^-
+   |                                       | |
+   |                                       | borrowed value does not live long enough
+   |                                       a temporary with access to the borrow is created here ...
 ...
-LL |     };
-   |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `MutexGuard`
-   |     |
-   |     `mutex` dropped here while still borrowed
+LL |     { let l = (); perform!(l) };
+   |           -       ----------- -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
+   |           |       |           |
+   |           |       |           `l` dropped here while still borrowed
+   |           |       in this macro invocation
+   |           binding `l` declared here
    |
+   = note: this error originates in the macro `perform` (in Nightly builds, run with -Z macro-backtrace for more info)
 help: consider adding semicolon after the expression so its temporaries are dropped sooner, before the local variables declared by the block are dropped
    |
-LL |         write!(Out, "{}", mutex.lock());
-   |                                        +
+LL |     { let l = (); perform!(l); };
+   |                              +
 
-error[E0597]: `mutex` does not live long enough
-  --> $DIR/span-semicolon-issue-139049.rs:48:41
+error[E0597]: `l` does not live long enough
+  --> $DIR/span-semicolon-issue-139049.rs:11:41
    |
-LL |         let mutex = Mutex;
-   |             ----- binding `mutex` declared here
-LL |         write!(std::io::stdout(), "{}", mutex.lock())
-   |                                         ^^^^^-------
-   |                                         |
-   |                                         borrowed value does not live long enough
-   |                                         a temporary with access to the borrow is created here ...
+LL | macro_rules! perform { ($e:expr) => { D(&$e).end() } }
+   |                                       --^^^-
+   |                                       | |
+   |                                       | borrowed value does not live long enough
+   |                                       a temporary with access to the borrow is created here ...
 ...
-LL |     };
-   |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `MutexGuard`
-   |     |
-   |     `mutex` dropped here while still borrowed
+LL |     let _x = { let l = (); perform!(l) };
+   |                    -       ----------- -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `D`
+   |                    |       |           |
+   |                    |       |           `l` dropped here while still borrowed
+   |                    |       in this macro invocation
+   |                    binding `l` declared here
    |
    = note: the temporary is part of an expression at the end of a block;
            consider forcing this temporary to be dropped sooner, before the block's local variables are dropped
+   = note: this error originates in the macro `perform` (in Nightly builds, run with -Z macro-backtrace for more info)
 help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
    |
-LL |         let x = write!(std::io::stdout(), "{}", mutex.lock()); x
-   |         +++++++                                              +++
+LL |     let _x = { let l = (); let x = perform!(l); x };
+   |                            +++++++            +++
 
 error: aborting due to 2 previous errors
 

